--- etc/rc.d/rc.S	2016-06-12 21:33:03.000000000 -0400
+++ etc/rc.d/rc.S.new	2020-11-08 18:18:33.325006268 -0500
@@ -123,6 +123,9 @@
 # ro -- this will cause --readonly to be passed to the cryptsetup program while
 # opening the LUKS volume.
 #
+# lvm -- this will cause a scan and activation of LVM volume groups after all
+# LUKS devices are opened.
+#
 # swap -- this option cannot be used with other options.  The device given will
 # be formatted as a new encrypted volume with a random key on boot, and used as
 # swap.
@@ -134,6 +137,8 @@
     # Try to load a device-mapper kernel module:
     /sbin/modprobe -q dm-mod
   fi
+  
+  LUKSLVM="no"  
   # NOTE: we only support LUKS formatted volumes (except for swap)!
   cat /etc/crypttab | grep -v "^#" | grep -v "^$" | while read line; do
     eval LUKSARRAY=( $line )
@@ -144,6 +149,7 @@
     LUKSOPTS=""
     if echo $OPTS | grep -wq ro ; then LUKSOPTS="${LUKSOPTS} --readonly" ; fi
     if echo $OPTS | grep -wq discard ; then LUKSOPTS="${LUKSOPTS} --allow-discards" ; fi
+    if echo $OPTS | grep -wq lvm ; then LUKSLVM="yes" ; fi    
     # Skip LUKS volumes that were already unlocked (in the initrd):
     /sbin/cryptsetup status $LUKS 2>/dev/null | head -n 1 | grep -q "is active" && continue
     if /sbin/cryptsetup isLuks $DEV 2>/dev/null ; then
@@ -172,6 +178,16 @@
       mkswap /dev/mapper/$LUKS
     fi
   done
+
+#  if [ "${LUKSLVM}" = "yes" ]; then
+      # Scan for new volume groups:
+      /sbin/vgscan --mknodes --ignorelockingfailure 2> /dev/null
+      if [ $? = 0 ]; then
+	  # Make volume groups available to the kernel.
+	  # This should also make logical volumes available.
+	  /sbin/vgchange -ay --ignorelockingfailure
+      fi
+#  fi
 fi
 
 # Enable swapping:
