Index: scid/scid/tcl/options.tcl
===================================================================
--- scid.orig/scid/tcl/options.tcl
+++ scid/scid/tcl/options.tcl
@@ -566,7 +566,7 @@ set ::utils::sound::announceForward 0
 set ::utils::sound::announceBack 0
 
 # Spell-checking file: default is "spelling.ssp".
-set spellCheckFile [file join $scidShareDir "spelling.ssp"]
+set spellCheckFile [file join $scidShareDir "data/spelling.ssp"]
 
 # book configuration
 set ::book::lastBook "" ; # book name without extension (.bin)
